src/bun.js/api/glob.zig

```diff
@@ -0,0 +1,455 @@
+        dot,
+        absolute,
+        follow_symlinks,
+        error_on_broken_symlinks,
+        only_files,
+    ) catch {
+        globalThis.throw("Out of memory", .{});
+        return null;
+    }) {
+        .err => |err| {
+            globalThis.throwValue(err.toJSC(globalThis));
+            return null;
+        },
+        else => {},
+    }
+
+    return globWalker;
+}
+
+pub fn constructor(
+    globalThis: *JSC.JSGlobalObject,
+    callframe: *JSC.CallFrame,
+) callconv(.C) ?*Glob {
+    const alloc = getAllocator(globalThis);
+
+    const arguments_ = callframe.arguments(1);
+    var arguments = JSC.Node.ArgumentsSlice.init(globalThis.bunVM(), arguments_.slice());
+    defer arguments.deinit();
+    const pat_arg = arguments.nextEat() orelse {
+        globalThis.throw("Glob.constructor: expected 1 arguments, got 0", .{});
+        return null;
+    };
+
+    if (!pat_arg.isString()) {
+        globalThis.throw("Glob.constructor: first argument is not a string", .{});
+        return null;
+    }
+
+    var pat_str: []u8 = pat_arg.getZigString(globalThis).toOwnedSlice(globalThis.bunVM().allocator) catch @panic("OOM");
+
+    const all_ascii = isAllAscii(pat_str);
+
+    var glob = alloc.create(Glob) catch @panic("OOM");
+    glob.* = .{ .pattern = pat_str, .is_ascii = all_ascii };
+
+    if (!all_ascii) {
+        var codepoints = std.ArrayList(u32).initCapacity(alloc, glob.pattern.len * 2) catch {
+            globalThis.throwOutOfMemory();
+            return null;
+        };
+        errdefer codepoints.deinit();
+
+        convertUtf8(&codepoints, glob.pattern) catch {
+            globalThis.throwOutOfMemory();
+            return null;
+        };
+
+        glob.pattern_codepoints = codepoints;
+    }
+
+    return glob;
+}
+
+pub fn finalize(
+    this: *Glob,
+) callconv(.C) void {
+    const alloc = JSC.VirtualMachine.get().allocator;
+    alloc.free(this.pattern);
+    alloc.destroy(this);
+}
+
+pub fn hasPendingActivity(this: *Glob) callconv(.C) bool {
+    @fence(.SeqCst);
+    return this.has_pending_activity.load(.SeqCst);
+}
+
+fn updateHasPendingActivityFlag(has_pending_activity: *std.atomic.Atomic(bool), value: bool) void {
+    @fence(.SeqCst);
+    has_pending_activity.store(value, .SeqCst);
+}
+
+pub fn __scan(this: *Glob, globalThis: *JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+    const alloc = getAllocator(globalThis);
+
+    const arguments_ = callframe.arguments(1);
+    var arguments = JSC.Node.ArgumentsSlice.init(globalThis.bunVM(), arguments_.slice());
+    defer arguments.deinit();
+
+    var arena = std.heap.ArenaAllocator.init(alloc);
+    var globWalker = this.makeGlobWalker(globalThis, &arguments, "scan", alloc, &arena) orelse {
+        arena.deinit();
+        return .undefined;
+    };
+
+    updateHasPendingActivityFlag(&this.has_pending_activity, true);
+    var task = WalkTask.create(globalThis, alloc, globWalker, &this.has_pending_activity) catch {
```

Comment:
what if there are multiple simultaneous walkers? should `has_pending_activity` be a number that gets incremented/decremented?